File: Backend/Nodejs/src/controllers/playlist.ts
import { Request, Response } from "express";
import { Playlist } from "../models/types";
import { PlaylistModel } from "../models/playlist";

export class PlaylistController {
  static createPlaylist(req: Request, res: Response) {
    const playlist = req.body as Playlist;
    const file = req.file as Express.MulterS3.File;
    if (!playlist || !file) {
      return res.status(400).json({ MESSAGE: "Faltan parámetros" });
    }
    PlaylistModel.createPlaylist(
      playlist,
      file,
      (response: any, ok: Boolean) => {
        // Respuesta
        res.status(ok ? 200 : 400).json(response);
      },
    );
  }

  static getUserPlaylists(req: Request, res: Response) {
    const { email } = req.params;
    if (!email) return res.status(401).json({ message: "Falta el correo" });
    PlaylistModel.getUserPlaylist(
      { email },
      (response: any, ok: Boolean) => {
        res.status(ok ? 200 : 400).json(response);
      },
    );
  }


  static getPlaylist(req: Request, res: Response) {
    const { id } = req.params
    if (!id) return res.status(401).json({ message: "Falta el id de la playlist" });
    
    PlaylistModel.getPlaylist({ id }, (response: any, ok: Boolean) => {
        res.status(ok ? 200 : 400).json(response);
      },);
  }


  static getSongs(req: Request, res: Response) {
    const { id } = req.params;
    if (!id) {
      return res.status(401).json({ message: "Falta el id de la playlist" });
    }
    PlaylistModel.getSongs({ id }, (response: any, ok: Boolean) => {
      res.status(ok ? 200 : 400).json(response);
    });
  }
  
  static getSongsNotInPlaylist(req: Request, res: Response) {
    const { id } = req.params;
    if (!id) {
      return res.status(401).json({ message: "Falta el id de la playlist" });
    }
    PlaylistModel.getSongsNotInPlaylist({ id }, (response: any, ok: Boolean) => {
      res.status(ok ? 200 : 400).json(response);
    });
  }

  static addSong(req: Request, res: Response) {
    const { playlist, song, email } = req.body;
    if (!playlist || !song || !email) {
      return res.status(401).json({ MESSAGE: "Faltan datos" });
    }
    PlaylistModel.addSong(playlist, song, email, (response, ok) => {
      res.status(ok ? 200 : 400).json(response);
    });
  }

  static removeSong(req: Request, res: Response) {
    const { playlist, song, email } = req.body;
    if (!playlist || !song || !email) {
      return res.status(401).json({ MESSAGE: "Faltan datos" });
    }
    PlaylistModel.removeSong(playlist, song, email, (response, ok) => {
      res.status(ok ? 200 : 400).json(response);
    });
  }

  static removePlaylist(req: Request, res: Response) {
    const { playlist, email } = req.body;
    if (!playlist || !email) {
      return res.status(401).json({ MESSAGE: "Faltan datos" });
    }
    const id = parseInt(playlist);
    PlaylistModel.deletePlaylist({ id, email }, (response, ok) => {
      res.status(ok ? 200 : 400).json(response);
    });
  }
}

File: Backend/Python/src/controllers/playlist.py
from flask import request
from src.libs.s3client import upload_file
from src.models.playlist import PlaylistModel


class PlaylistController:

    @staticmethod
    def create_playlist():
        body = request.form
        file = request.files
        if len(body) == 0 or len(file) == 0:
            return {"MESSAGE": "Faltan datos", "TYPE": "ERROR"}, 401
        cover_location = upload_file("playlist", file["cover"])
        if cover_location == "":
            return {"MESSAGE": "No se pudo subir la foto", "TYPE": "ERROR"}, 401
        response = PlaylistModel.create_playlist(body, cover_location)
        return response[0], (200 if response[1] else 400)

    @staticmethod
    def get_users_playlists(email):
        response = PlaylistModel.get_users_playlists(email)
        return response[0], (200 if response[1] else 400)

    @staticmethod
    def get_songs(id):
        response = PlaylistModel.get_songs(id)
        return response[0], (200 if response[1] else 400)

    @staticmethod
    def add_song():
        body = request.json
        if "email" not in body or "playlist" not in body or "song" not in body:
            return {"MESSAGE": "Faltan datos", "TYPE": "ERROR"}, 401
        response = PlaylistModel.add_song(body["playlist"], body["song"], body["email"])
        return response[0], (200 if response[1] else 400)

    @staticmethod
    def remove_song():
        body = request.json
        if "email" not in body or "playlist" not in body or "song" not in body:
            return {"MESSAGE": "Faltan datos", "TYPE": "ERROR"}, 401
        response = PlaylistModel.remove_song(
            body["playlist"], body["song"], body["email"]
        )
        return response[0], (200 if response[1] else 400)

    @staticmethod
    def remove_playlist():
        body = request.json
        if "email" not in body or "playlist" not in body:
            return {"MESSAGE": "Faltan datos", "TYPE": "ERROR"}, 401
        response = PlaylistModel.remove_playlist(body["playlist"], body["email"])
        return response[0], (200 if response[1] else 400)

    @staticmethod
    def edit_playlist():
        body = dict(request.form)
        file = request.files
        if len(body) == 0:
            return {"MESSAGE": "Faltan datos", "TYPE": "ERROR"}, 401
        
        cover_location = ""
        if "cover" in file:
            cover_location = upload_file("playlist", file["cover"])
            
        response = PlaylistModel.edit_playlist(body, cover_location)
        return response[0], (200 if response[1] else 400)

    @staticmethod
    def get_song_not_playlist(id):
        response = PlaylistModel.songs_not_playlist(id)
        return response[0], (200 if response[1] else 400)

    @staticmethod
    def get_playlist(id):
        response = PlaylistModel.get_playlist(id)
        return response[0], (200 if response[1] else 400)

